File: app/src/main/java/com/neversoft/spotlight/search/SearchEngine.kt
package com.neversoft.spotlight.search

import android.content.Context
import com.neversoft.spotlight.model.FilterSelection
import com.neversoft.spotlight.model.PrimaryFilter
import com.neversoft.spotlight.model.ResultType
import com.neversoft.spotlight.model.SearchResult
import com.neversoft.spotlight.model.StorageScope
import com.neversoft.spotlight.model.SubFilter
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.isActive
import kotlinx.coroutines.withContext
import java.util.Locale

/**
 * Aggregates every provider into one device-wide search, routed by the active
 * filter selection. Runs on [Dispatchers.IO] and is fully cancellable so each
 * new keystroke can abandon the previous query.
 */
class SearchEngine(context: Context) {

    private val apps = AppSearchProvider(context)
    private val media = MediaSearchProvider(context)
    private val files = FileSearchProvider(context)
    private val contacts = ContactSearchProvider(context)

    fun hasFullFileAccess(): Boolean = files.hasFullAccess()

    suspend fun search(
        query: String,
        selection: FilterSelection,
        scope: StorageScope = StorageScope.ALL,
        limit: Int = DEFAULT_LIMIT,
    ): List<SearchResult> = withContext(Dispatchers.IO) {
        val q = query.trim()
        if (q.isEmpty()) return@withContext emptyList()

        val active = { isActive }
        val deadline = System.currentTimeMillis() + WALK_BUDGET_MS
        val results = ArrayList<SearchResult>()

        when (selection.primary) {
            PrimaryFilter.ALL -> {
                results += apps.search(q, 40)
                ensureActive()
                results += contacts.search(q, 30, active)
                ensureActive()
                results += media.search(
                    q, setOf(ResultType.IMAGE, ResultType.VIDEO, ResultType.AUDIO), 90, active, scope,
                )
                ensureActive()
                results += files.search(
                    q,
                    FileSearchSpec(includeFiles = true, includeFolders = true),
                    160, deadline, active, scope,
                )
            }

            PrimaryFilter.MEDIA -> {
                val types = when (selection.sub) {
                    SubFilter.PHOTOS -> setOf(ResultType.IMAGE)
                    SubFilter.VIDEOS -> setOf(ResultType.VIDEO)
                    SubFilter.AUDIO -> setOf(ResultType.AUDIO)
                    else -> setOf(ResultType.IMAGE, ResultType.VIDEO, ResultType.AUDIO)
                }
                results += media.search(q, types, limit, active, scope)
            }

            PrimaryFilter.FILES -> {
                val spec = when (selection.sub) {
                    SubFilter.DOCUMENTS -> FileSearchSpec(extensions = FileTypes.DOCUMENT)
                    SubFilter.DOWNLOADS -> FileSearchSpec(downloadsOnly = true)
                    SubFilter.ARCHIVES -> FileSearchSpec(extensions = FileTypes.ARCHIVE)
                    SubFilter.APKS -> FileSearchSpec(extensions = FileTypes.APK)
                    else -> FileSearchSpec(includeFiles = true, includeFolders = true)
                }
                results += files.search(q, spec, limit, deadline, active, scope)
            }

            PrimaryFilter.HIDDEN -> {
                val spec = when (selection.sub) {
                    SubFilter.HIDDEN_FOLDERS ->
                        FileSearchSpec(includeFiles = false, includeFolders = true, hiddenOnly = true)
                    SubFilter.HIDDEN_FILES ->
                        FileSearchSpec(includeFiles = true, includeFolders = false, hiddenOnly = true)
                    else ->
                        FileSearchSpec(includeFiles = true, includeFolders = true, hiddenOnly = true)
                }
                results += files.search(q, spec, limit, deadline, active, scope)
            }
        }

        ensureActive()
        rankAndDedupe(results, q, limit)
    }

    private fun rankAndDedupe(list: List<SearchResult>, query: String, limit: Int): List<SearchResult> {
        val lower = query.lowercase(Locale.getDefault())
        return list.distinctBy { it.id }
            .sortedWith(
                compareBy(
                    { matchScore(it.title, lower) },
                    { typeOrder(it.type) },
                    { -it.dateModified },
                ),
            )
            .take(limit)
    }

    /** Lower is better: exact match first, then prefix, then substring. */
    private fun matchScore(title: String, lowerQuery: String): Int {
        val t = title.lowercase(Locale.getDefault())
        return when {
            t == lowerQuery -> 0
            t.startsWith(lowerQuery) -> 1
            else -> 2
        }
    }

    /** Surface apps & contacts above raw files in mixed (ALL) results. */
    private fun typeOrder(type: ResultType): Int = when (type) {
        ResultType.APP -> 0
        ResultType.CONTACT -> 1
        ResultType.IMAGE, ResultType.VIDEO, ResultType.AUDIO -> 2
        ResultType.DOCUMENT, ResultType.APK, ResultType.ARCHIVE -> 3
        ResultType.FOLDER -> 4
        ResultType.FILE -> 5
    }

    companion object {
        const val DEFAULT_LIMIT = 400
        const val WALK_BUDGET_MS = 2500L
    }
}

File: windows/Spotlight/Search/SearchEngine.cs
using NeverSoft.Spotlight.Models;
using NeverSoft.Spotlight.Services;

namespace NeverSoft.Spotlight.Search;

/// <summary>
/// Aggregates every provider into one device-wide search, routed by the active filter
/// selection — a faithful port of the Android SearchEngine. Each filter only queries the
/// providers it needs (so "Media" never walks for documents, "Files" never lists apps, and
/// so on), then results are ranked (exact → prefix → substring), personalized by usage, and
/// de-duplicated. Runs off the UI thread and is fully cancellable so each keystroke can
/// abandon the previous query.
/// </summary>
public sealed class SearchEngine
{
    public const int DefaultLimit = 400;
    public const long WalkBudgetMs = 2500;

    private readonly AppSearchProvider _apps = new();
    private readonly ContactSearchProvider _contacts = new();
    private readonly MediaSearchProvider _media;
    private readonly FileSearchProvider _files;
    private readonly UsageStore _usage;

    public SearchEngine(UsageStore usage, FileIndex index)
    {
        _usage = usage;
        _media = new MediaSearchProvider(index);
        _files = new FileSearchProvider(index);
    }

    public Task<IReadOnlyList<SearchResult>> SearchAsync(
        string query, FilterSelection selection, int limit = DefaultLimit, CancellationToken ct = default)
        => Task.Run(() => Search(query, selection, limit, ct), ct);

    private IReadOnlyList<SearchResult> Search(
        string query, FilterSelection selection, int limit, CancellationToken ct)
    {
        // An empty query is a valid "browse" request: every provider treats it as match-all,
        // so the list populates from the active filter before the user types anything.
        var q = query.Trim();
        long deadline = Environment.TickCount64 + WalkBudgetMs;
        var results = new List<SearchResult>();

        switch (selection.Primary)
        {
            case PrimaryFilter.All:
                results.AddRange(_apps.Search(q, 40, ct));
                ct.ThrowIfCancellationRequested();
                results.AddRange(_contacts.Search(q, 30, ct));
                ct.ThrowIfCancellationRequested();
                results.AddRange(_media.Search(
                    q, new[] { ResultType.Image, ResultType.Video, ResultType.Audio }, 90, deadline, ct));
                ct.ThrowIfCancellationRequested();
                results.AddRange(_files.Search(
                    q, new FileSearchSpec(IncludeFiles: true, IncludeFolders: true), 160, deadline, ct));
                break;

            case PrimaryFilter.Media:
                var types = selection.Sub switch
                {
                    SubFilter.Photos => new[] { ResultType.Image },
                    SubFilter.Videos => new[] { ResultType.Video },
                    SubFilter.Audio => new[] { ResultType.Audio },
                    _ => new[] { ResultType.Image, ResultType.Video, ResultType.Audio },
                };
                results.AddRange(_media.Search(q, types, limit, deadline, ct));
                break;

            case PrimaryFilter.Files:
                var spec = selection.Sub switch
                {
                    SubFilter.Documents => new FileSearchSpec(Extensions: FileTypes.Document),
                    SubFilter.Downloads => new FileSearchSpec(DownloadsOnly: true),
                    SubFilter.Archives => new FileSearchSpec(Extensions: FileTypes.Archive),
                    SubFilter.Programs => new FileSearchSpec(Extensions: FileTypes.Program),
                    _ => new FileSearchSpec(IncludeFiles: true, IncludeFolders: true),
                };
                results.AddRange(_files.Search(q, spec, limit, deadline, ct));
                break;

            case PrimaryFilter.Hidden:
                var hidden = selection.Sub switch
                {
                    SubFilter.HiddenFolders =>
                        new FileSearchSpec(IncludeFiles: false, IncludeFolders: true, HiddenOnly: true),
                    SubFilter.HiddenFiles =>
                        new FileSearchSpec(IncludeFiles: true, IncludeFolders: false, HiddenOnly: true),
                    _ =>
                        new FileSearchSpec(IncludeFiles: true, IncludeFolders: true, HiddenOnly: true),
                };
                results.AddRange(_files.Search(q, hidden, limit, deadline, ct));
                break;
        }

        ct.ThrowIfCancellationRequested();
        return RankAndDedupe(results, q, limit);
    }

    private IReadOnlyList<SearchResult> RankAndDedupe(List<SearchResult> list, string query, int limit)
    {
        var lower = query.ToLowerInvariant();
        return list
            .GroupBy(r => r.Id)
            .Select(g => g.First())                       // distinctBy id, preserving order
            .OrderBy(r => MatchScore(r.Title, lower))     // exact → prefix → substring
            .ThenBy(r => TypeOrder(r.Type))               // apps & contacts float up in mixed results
            .ThenByDescending(r => _usage.Score(r.Id))    // luxury: personalize by how often you launch it
            .ThenByDescending(r => r.DateModified)        // newest first
            .Take(limit)
            .ToList();
    }

    /// <summary>Lower is better: exact match first, then prefix, then substring.</summary>
    private static int MatchScore(string title, string lowerQuery)
    {
        var t = title.ToLowerInvariant();
        if (t == lowerQuery) return 0;
        if (t.StartsWith(lowerQuery, StringComparison.Ordinal)) return 1;
        return 2;
    }

    /// <summary>Surface apps & contacts above raw files in mixed (All) results.</summary>
    private static int TypeOrder(ResultType type) => type switch
    {
        ResultType.App => 0,
        ResultType.Contact => 1,
        ResultType.Image or ResultType.Video or ResultType.Audio => 2,
        ResultType.Document or ResultType.Program or ResultType.Archive => 3,
        ResultType.Folder => 4,
        ResultType.File => 5,
        _ => 6,
    };
}
